### ISSUE TITLE
DISTINCTCOUNT and DISTINCTCOUNTHLL aggregations are really slow

### ISSUE THREAD

**Issue by buchireddy** [apache/incubator-pinot#5907]
Pinot version used is `0.4.0`

Issue: DISTINCTCOUNT on a string column is much much slower than the group by on that same column.
**DISTINCTCOUNT query:**
```sql
Select DISTINCTCOUNT(foo) FROM myTable WHERE col1 = 'value1' AND time_col >= 1596750994918 AND time_col < 1597960594918 AND foo != 'null' limit 1
```

**group by query**
```sql
Select foo, COUNT(*) FROM myTable WHERE col1 = 'value1' AND time_col >= 1596750994918 AND time_col < 1597960594918 AND foo != 'null' group by foo limit 10000
```

We don't have an index on column `foo`. The DISTINCTCOUNT query was taking anywhere b/w 3 and 6s while the group by query returns in <500ms most of the time. DISTINCTCOUNTHLL was equally slower, if not more in some cases, than DISTINCTCOUNT query.

cc @kishoreg

**Comment by Jackie-Jiang** [comment]
@buchireddy Thanks for reporting the issue. Could you please share more stats of the query? How many documents does this query scan? How many distinct values does it have?
The performance of these 3 queries should be: `DISTINCTCOUNTHLL` > `DISTINCTCOUNT` > Group-by, unless only partial groups are returned.

**Comment by buchireddy** [comment]
> @buchireddy Thanks for reporting the issue. Could you please share more stats of the query? How many documents does this query scan? How many distinct values does it have?
> The performance of these 3 queries should be: `DISTINCTCOUNTHLL` > `DISTINCTCOUNT` > Group-by, unless only partial groups are returned.

```json
{
    "resultTable": {
        "dataSchema": {
            "columnDataTypes": ["INT"],
            "columnNames": ["distinctcount(service_id)"]
        },
        "rows": [
            [2]
        ]
    },
    "exceptions": [],
    "numServersQueried": 4,
    "numServersResponded": 4,
    "numSegmentsQueried": 180,
    "numSegmentsProcessed": 180,
    "numSegmentsMatched": 179,
    "numConsumingSegmentsQueried": 8,
    "numDocsScanned": 127901110,
    "numEntriesScannedInFilter": 131526370,
    "numEntriesScannedPostFilter": 127901110,
    "numGroupsLimitReached": false,
    "totalDocs": 128941076,
    "timeUsedMs": 3339,
    "segmentStatistics": [],
    "traceInfo": {},
    "minConsumingFreshnessTimeMs": 1597961648907
}
```

**Comment by kishoreg** [comment]
@Jackie-Jiang 500ms (group by) vs 3-6sec (distinct) is a big difference when there are only 2 distinct values

**Comment by kishoreg** [comment]
one reason might be that group by was using dictionary whereas distinct is not. I think you have fixed this in the recent version.

**Comment by Jackie-Jiang** [comment]
@kishoreg Good point, especially when the column is `STRING`.
@buchireddy There is an optimization on `DISTINCTCOUNT` recently, which should make it always faster than the group-by. You might want to try it using the current master

**Comment by buchireddy** [comment]
@Jackie-Jiang do you know if that fix is present in `0.5.0` or after that?

**Comment by Jackie-Jiang** [comment]
@buchireddy It will be included in the `0.5.0` release

**Comment by kishoreg** [comment]
@buchireddy please update this ticket once you have the new version
